### DISCUSSION TITLE
Is auto binding a feature?

### DISCUSSION THREAD

**Discussion by JordanMarr** [ninject/Ninject#403] [General]
I am using Ninject with a .NET Framework 4.8 project, and I just noticed that it seems to be auto-registering things.
This is not a problem (it's actually very convenient); however, I was wondering if there is any kind of documentation or mention of this as a feature.

**Comment by lord-executor** [comment]
I'm assuming that you are referring to the fact that you can create an instance of basically _any_ class with Ninject without defining a binding for it. For example, assuming that `MyClass` is a somewhat normal class `Kernel.Get<MyClass>()` just works without having to define a binding like `Kernel.Bind<MyClass>().ToSelf()`.

That is indeed a _feature_, although one that I have grown to dislike 😝. I call it "automatic implicit self bindings" and it originates in the [SelfBindingResolver](https://github.com/ninject/Ninject/blob/3.3.6/src/Ninject/Planning/Bindings/Resolvers/SelfBindingResolver.cs) class which is automatically added to the Ninject kernel instance. It is documented here https://github.com/ninject/Ninject/wiki/Building-an-IMissingBindingResolver-Component .

While I can definitely see the utility in automatic self bindings, I have also seen that it can cause some issues that are very hard to track down if somebody accidentally injects a concrete class instead of the associated interface as they are supposed to. The fact that it just "works" does not imply that it works _correctly_ since the interface binding can have sophisticated logic that will be completely side-stepped when the concrete instance is injected directly. In any case, while it isn't currently not configurable through the `INinjectSettings` (which I would like) it is still rather easy to get rid of, by just doing something like `Kernel.Components.Remove<IMissingBindingResolver, SelfBindingResolver>();`

**Reply by JordanMarr** [reply]
Thank you for the great explanation. I can totally see how it could be a foot gun, despite the initial convenience.
